onnx/defs/tensor/defs.cc

```diff
@@ -2370,31 +2370,40 @@ ONNX_OPERATOR_SET_SCHEMA(
-using a specified interpolation method (the mode) and a padding mode (for grid positions falling outside the 2-dimensional image).
-
-Elements in `grid[N, H_out, W_out]` are size-2 vectors specifying positions in the 2-dimensional space of `X`.
-They are used to interpolate output values of `Y[N, C, H_out, W_out]`.
-
-The GridSample operator is often used in doing grid generator and sampler in the [Spatial Transformer Networks](https://arxiv.org/abs/1506.02025).
-See also in [torch.nn.functional.grid_sample](https://pytorch.org/docs/master/generated/torch.nn.functional.grid_sample.html#torch-nn-functional-grid-sample).
+static const char* GridSample_ver20_doc = R"DOC(
+Given an input X and a flow-field grid, computes the output Y using X values and pixel locations from the grid.
+For spatial input X with shape (N, C, H, W), the grid will have shape (N, H_out, W_out, 2),
+the output Y will have shape (N, C, H_out, W_out). For volumetric input X with shape (N, C, D, H, W),
+the grid will have shape (N, D_out, H_out, W_out, 3), the output Y will have shape (N, C, D_out, H_out, W_out).
+More generally, for an input X of rank r+2 with shape (N, C, d1, d2, …, dr),
+the grid will have shape (N, D1_out, D2_out, …, Dr_out, r), the output Y will have shape (N, C, D1_out, D2_out, …, Dr_out).
+
+The tensor X contains values at centers of square pixels (voxels, etc) locations such as (n, c, d1_in, d2_in, …, dr_in).
+The (n, d1_out, d2_out, …, dr_out, :) values from the tensor grid are the normalized positions for interpolating the values
+at the (n, c, d1_out, d2_out, …, dr_out) locations from the output tensor Y using a specified interpolation method (the mode)
+and a padding mode (for grid positions falling outside the 2-dimensional image).
+
+For example, the values in grid[n, h_out, w_out, :] are size-2 vectors specifying normalized positions in the 2-dimensional space of X.
+They are used to interpolate output values of Y[n, c, h_out, w_out].
+
+The GridSample operator is often used in doing grid generator and sampler in the
+[Spatial Transformer Networks](https://arxiv.org/abs/1506.02025).
+See also in [torch.nn.functional.grid_sample](https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html).
 )DOC";
 
 ONNX_OPERATOR_SET_SCHEMA(
     GridSample,
-    16,
+    20,
     OpSchema()
         .Attr(
             "mode",
-            "Three interpolation modes: bilinear (default), nearest and bicubic.",
+            "Three interpolation modes: linear (default), nearest and cubic. "
+            "The \"linear\" mode includes linear and N-linear interpolation modes depending on the number of spatial dimensions "
+            "of the input tensor (i.e. linear for 1 spatial dimension, bilinear for 2 spatial dimensions, etc.). "
+            "The \"cubic\" mode also includes N-cubic interpolation modes following the same rules. The \"nearest\" mode rounds "
```

Comment:
I don't know much about how the cubic interpolation works. Is it unambiguously defined for general N-dimensions? Is there any pointer or link to the definition of cubic interpolation? Does it take 4 points (in the case of 1D) to do the interpolation, 2 points before and 2 points after the target point?

Reply:
The cubic interpolation that the community commonly used is the [cubic convolution algorithm](https://en.wikipedia.org/wiki/Bicubic_interpolation#Bicubic_convolution_algorithm).
In fact, ONNX [Reisze](https://github.com/onnx/onnx/blob/v1.13.1/onnx/reference/ops/op_resize.py) cubic interpolation which also works for any arbitrary dimension uses this algorithm. 
You are correct. It does take 4 points per dimension for interpolation. 2 before the coordinate and 2 after the coordinate along the specific dimension.